Python updates 2022-02-15

mweinelt opened the pull request:
This cycle is about creating a cheap baseline evaluation for the expected setuptools breakage to have a comparison against master.

https://hydra.nixos.org/eval/1744066

In the meantime it has been rebased and retargeted onto staging.

This is closely related to https://github.com/NixOS/nixpkgs/issues/155516 and we will start updating or removing packages that rely on this feature.

Previous python-unstable: #154969

mweinelt committed:
python3Packages.setuptools: 57.2.0 -> 60.8.2

mweinelt commented:
Basic evals checks fail in pytest.

```
_____________________________ test_version_verbose _____________________________

pytester = <Pytester PosixPath('/build/pytest-of-nixbld/pytest-0/test_version_verbose0')>
pytestconfig = <_pytest.config.Config object at 0x7ffff6a56580>
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7ffff39d9ca0>

    def test_version_verbose(pytester: Pytester, pytestconfig, monkeypatch) -> None:
        monkeypatch.delenv("PYTEST_DISABLE_PLUGIN_AUTOLOAD")
        result = pytester.runpytest("--version", "--version")
        assert result.ret == 0
        result.stderr.fnmatch_lines([f"*pytest*{pytest.__version__}*imported from*"])
        if pytestconfig.pluginmanager.list_plugin_distinfo():
>           result.stderr.fnmatch_lines(["*setuptools registered plugins:", "*at*"])
E           Failed: nomatch: '*setuptools registered plugins:'
E               and: 'This is pytest version 6.2.5, imported from /nix/store/jfn7276vsd1l8daa1g14g5qdhdjzn419-python3.9-pytest-6.2.5/lib/python3.9/site-packages/pytest/__init__.py'
E           remains unmatched: '*setuptools registered plugins:'

/build/pytest-6.2.5/testing/test_helpconfig.py:12: Failed
----------------------------- Captured stderr call -----------------------------
This is pytest version 6.2.5, imported from /nix/store/jfn7276vsd1l8daa1g14g5qdhdjzn419-python3.9-pytest-6.2.5/lib/python3.9/site-packages/pytest/__init__.py
=========================== short test summary info ============================
FAILED testing/test_helpconfig.py::test_version_verbose - Failed: nomatch: '*...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
```

dotlambda commented:
Where did the pip update go?

mweinelt commented:
Force pushed it away, it wasn't really relevant and can be added back again later.

mweinelt committed:
python3Packages.parse-type: 0.5.6 -> 0.6.0

https://github.com/jenisys/parse_type/releases/tag/v0.6.0

mweinelt committed:
python3Packages.readme_renderer: disable tests that rely on old distutils behaviour

mweinelt committed:
python3Packages.pybind11: disable tests that parse setuptools output

mweinelt committed:
python3Packages.async-lru: fix tests

mweinelt committed:
python3Packages.bitcoin-price-api: drop

mweinelt committed:
python3Packages.django-widget-tweaks: fix build and update tests

mweinelt committed:
python3Packages.hypothesmith: patch lark requirement

The library was renamed and this package didn't follow … yet?

mweinelt committed:
python3Packages.hypothesmith: configure parallel build

Including pytest-xdist without configuring it has no effect...

mweinelt committed:
isso: fix check phase

Commit:
python3Packages.portalocker: disable setuptools sensitive test

Commit:
python3Packages.keepalive: mark brokenk for setuptools>=58

Commit:
python3Packages.parameterizedtestcase: mark brokenk for setuptools>=58

jonringer commented:
`pyownet` is no longer compatible

Commit:
azure-cli: fix build

SuperSandro2000 commented on pkgs/development/python-modules/django-widget-tweaks/default.nix:
```diff
@@ -1,4 +1,16 @@
-{ buildPythonPackage, fetchFromGitHub, python, lib, django }:
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+
+# native
+, setuptools-scm
+
+# propagated
+, django
+
+# tests
```
```suggestion
, fetchFromGitHub
, setuptools-scm
, django
```

piegamesde replied:
Are you really suggesting to *remove* code comments? Why?

SuperSandro2000 replied:
> Are you really suggesting to remove code comments? Why?

We are not removing any useful comments here, we are not adding unhelpful ones.
They only duplicate information which can already be easily looked at a few lines below and create new questions like what happens if something is used in native, propagatedBuildInputs and tests. Also they add new meta data which needs to be maintained and if bad maintained makes the code harder to understand.

SuperSandro2000 commented on pkgs/development/python-modules/parameterizedtestcase/default.nix:
```diff
@@ -27,5 +27,6 @@ buildPythonPackage rec {
     homepage = "https://github.com/msabramo/python_unittest_parameterized_test_case";
     license = licenses.mit;
     maintainers = with maintainers; [ dotlambda ];
+    broken = python.isPy3k; # uses use_2to3
```
Since python2 moved out this would be never false. Why not delete it?

SuperSandro2000 commented on pkgs/development/python-modules/parse-type/default.nix:
```diff
@@ -29,7 +29,7 @@ buildPythonPackage rec {
     substituteInPlace pytest.ini \
       --replace "--metadata PACKAGE_UNDER_TEST parse_type" "" \
-      --replace "--metadata PACKAGE_VERSION 0.5.6" "" \
+      --replace "--metadata PACKAGE_VERSION 0.6.0" "" \
```
```suggestion
      --replace "--metadata PACKAGE_VERSION ${version}" "" \
```

mweinelt committed:
python3Packages.nose-cover3: drop

risicle commented:
On macos 10.15:

```
python/_brotli.cc:5:10: fatal error: 'vector' file not found
#include <vector>
         ^~~~~~~~
1 error generated.
error: command '/nix/store/18i795dcmzx2jlc7560nf65qybp0y41p-clang-wrapper-11.1.0/bin/clang' failed with exit code 1
builder for '/nix/store/nrwlr2r5j85w2armqm1qhbabbmckipzw-python3.9-brotli-1.0.9.drv' failed with exit code 1
```

appears to be caused by the setuptools bump.

risicle commented:
On macos 10.15 `numpy` also fails to build:

```
...
...
  File "/private/tmp/nix-build-python3.9-numpy-1.21.5.drv-0/numpy-1.21.5/numpy/distutils/unixccompiler.py", line 52, in UnixCCompiler__compile
    display = '%s: %s' % (os.path.basename(self.compiler_so_cxx[0]), src)
AttributeError: 'UnixCCompiler' object has no attribute 'compiler_so_cxx'
```

risicle commented:
Ok think I have a fix for brotli/protobuf - will push

risicle commented:
It looks like this `SETUPTOOLS_USE_DISTUTILS="stdlib"` fix also works for the numpy failure, but it seems like it'll be needed for *all* packages that use the numpy distutils. So I'm thinking it would be a better idea to come up with an alternative to https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/numpy/numpy-distutils-C%2B%2B.patch that works with setuptools 60's bundled distutils.

risicle commented:
Oh it occurs to me that what's happening is setuptools 60 going back to a builtin distutils means it's using a distutils without https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/cpython/3.7/python-3.x-distutils-C++.patch

risicle commented:
I've mangled a version of the distutils-cpp patch to apply to the new setuptools and it seems to work. Will tidy and post tomorrow.

risicle committed:
python3Packages.setuptools: add patch for bundled distutils c++ support

setuptools 60.x (re-)introduces a bundled distutils, which we need
to patch with C++ support in the same way as we do with the stdlib's
distutils (mostly to placate clang on macos)

risicle commented:
Have built quite a lot of packages with this on macos 10.15 and am now only getting pre-existing failures.

mweinelt committed:
release-python.nix: add buildcatrust to tested set

This is the tool that manages our certificate store.

mweinelt commented:
The state of things right now is looking pretty good.
https://shells.darmstadt.ccc.de/~hexa/1744686.html

mweinelt commented:
Thanks @risicle for coming up with that patch. That was a big chunk of the heavy-lifting to get this PR merged!

trofi commented:
Bisect claims that 5c09870c0244bbcf47a84f379e05bf10c7aa3f0d `python3Packages.setuptools: 57.2.0 -> 60.8.2` broke tests for `python3Packages.scikit-build` in `staging`:

```
$ nix build -f. python3Packages.scikit-build -L
...
python3.9-scikit-build> tests/test_command_line.py::test_hide_listing[True-sdist] FAILED         [ 26%]
python3.9-scikit-build> tests/test_command_line.py::test_hide_listing[True-bdist_wheel] FAILED   [ 27%]
...
python3.9-scikit-build> =================================== FAILURES ===================================
python3.9-scikit-build> ________________________ test_hide_listing[True-sdist] _________________________
python3.9-scikit-build> action = 'sdist', hide_listing = True
python3.9-scikit-build> capfd = <_pytest.capture.CaptureFixture object at 0x7ffff495b1f0>
python3.9-scikit-build>     @pytest.mark.parametrize("action", ['sdist', 'bdist_wheel'])
python3.9-scikit-build>     @pytest.mark.parametrize("hide_listing", [True, False])
python3.9-scikit-build>     def test_hide_listing(action, hide_listing, capfd):
python3.9-scikit-build>         cmd = [action]
python3.9-scikit-build>         if hide_listing:
python3.9-scikit-build>             cmd.insert(0, "--hide-listing")
python3.9-scikit-build>         @project_setup_py_test("test-hide-listing", cmd, verbose_git=False, disable_languages_test=True)
python3.9-scikit-build>         def run():
python3.9-scikit-build>             pass
python3.9-scikit-build>         run()
python3.9-scikit-build>         out, _ = capfd.readouterr()
python3.9-scikit-build>         if hide_listing:
python3.9-scikit-build> >           assert to_platform_path("bonjour/__init__.py") not in out
python3.9-scikit-build> E           AssertionError: assert 'bonjour/__init__.py' not in 'running sdi... under it)\n'
python3.9-scikit-build> E             'bonjour/__init__.py' is contained here:
python3.9-scikit-build> E               running sdist
python3.9-scikit-build> E               running generate_source_manifest
python3.9-scikit-build> E               running check
python3.9-scikit-build> E               reading manifest file 'MANIFEST'
python3.9-scikit-build> E               creating hello-1.2.3
python3.9-scikit-build> E               creating hello-1.2.3/bonjour...
python3.9-scikit-build> E
python3.9-scikit-build> E             ...Full output truncated (19 lines hidden), use '-vv' to show
python3.9-scikit-build> _          = "hint: Using 'master' as the name for the initial branch. This default branch name\nhint: is subject to change. To con...d meta-data: url\n\nwarning: check: missing meta-data: if 'author' supplied, 'author_email' should be supplied too\n\n"
python3.9-scikit-build> action     = 'sdist'
python3.9-scikit-build> capfd      = <_pytest.capture.CaptureFixture object at 0x7ffff495b1f0>
python3.9-scikit-build> cmd        = ['--hide-listing', 'sdist']
python3.9-scikit-build> hide_listing = True
python3.9-scikit-build> out        = "running sdist\nrunning generate_source_manifest\nrunning check\nreading manifest file 'MANIFEST'\ncreating hello-1.2....nd adding 'hello-1.2.3' to it\ncreating dist\nCreating tar archive\nremoving 'hello-1.2.3' (and everything under it)\n"
python3.9-scikit-build> run        = <function test_hide_listing.<locals>.run at 0x7ffff46da310>
python3.9-scikit-build> tests/test_command_line.py:176: AssertionError
...
python3.9-scikit-build> FAILED tests/test_command_line.py::test_hide_listing[True-sdist] - AssertionE...
python3.9-scikit-build> FAILED tests/test_command_line.py::test_hide_listing[True-bdist_wheel] - asse...
```

risicle commented:
https://github.com/scikit-build/scikit-build/pull/623

Looks like we're quite behind on scikit-build anyway.

risicle commented:
In fact the next "Python updates" includes a bump that may well fix it https://github.com/NixOS/nixpkgs/pull/162538/commits/ada1d706ba81aa7d47a666508be2bd36bd41dcf2

mweinelt commented:
I see two different tests failing with that update, maybe we should just disable those.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/build-platform-cpython-headers-instead-of-host-platform-headers-used-whenever-a-python-extension-is-cross-compiled/20096/4
